File: algo/depthFirstSearch.py
class Node:
    def __init__(self, name):
        self.children = []
        self.name = name

    def addChild(self, name):
        self.children.append(Node(name))
        return self

    def depthFirstSearch(self, array):
		# Write your code here.
		#array.append(self.name)
		array.append(self.name)

		for child in self.children:
			child.depthFirstSearch(array)

		return array


tests = [

    {
      "graph": {
        "nodes": [
          {"children": ["B", "C", "D"], "id": "A", "value": "A"},
          {"children": ["E", "F"], "id": "B", "value": "B"},
          {"children": [], "id": "C", "value": "C"},
          {"children": ["G", "H"], "id": "D", "value": "D"},
          {"children": [], "id": "E", "value": "E"},
          {"children": ["I", "J"], "id": "F", "value": "F"},
          {"children": ["K"], "id": "G", "value": "G"},
          {"children": [], "id": "H", "value": "H"},
          {"children": [], "id": "I", "value": "I"},
          {"children": [], "id": "J", "value": "J"},
          {"children": [], "id": "K", "value": "K"}
        ],
        "startNode": "A"
      },
    "expect":["A", "B", "E", "F", "I", "J", "C", "D", "G", "K", "H"]
    },

    {
      "graph": {
        "nodes": [
          {"children": ["B", "C"], "id": "A", "value": "A"},
          {"children": ["D"], "id": "B", "value": "B"},
          {"children": [], "id": "C", "value": "C"},
          {"children": [], "id": "D", "value": "D"}
        ],
        "startNode": "A"
      },
      "expect": ["A", "B", "D", "C"]
    }
]
idx=0
for case in tests:
    idx+=1
    print ("\n%s test %i *********" % ("*"*20 , idx))

    nodes={}
    for node in case["graph"]["nodes"]:
        nodes[node["id"]]=Node(node["value"])
        print("Created node %s" % node["id"])
    for node in case["graph"]["nodes"]:
        for child in node["children"]:
            nodes[ node["id"] ].children.append(nodes[child])
            print("Added child %s to %s " % (child,node["id"]))
    print("-- BST node list")
    startNode=nodes[case["graph"]["startNode"]]
    array=[]
    print(startNode.depthFirstSearch(array))
    print("-- Expected ")
    print(case["expect"])

File: InterviewTraining/go/problems/depthFirstSearch.go
//go:build problem

package main

/*
Original JS: js/depthFirstSearch.js

This file defines a Node type with a DepthFirstSearch method.
The runner expects a constructor-style symbol `New(value interface{}) interface{}`
that returns a *Node so it can instantiate nodes during tests.
*/

type Node struct {
	Name     string
	Children []*Node
}

func (n *Node) AddChild(child *Node) *Node {
	n.Children = append(n.Children, child)
	return n
}

// DepthFirstSearch performs DFS and returns a slice of visited names.
// TODO: implement traversal logic.
func (n *Node) DepthFirstSearch(arr []interface{}) []interface{} {
	return nil
}

// New constructs a Node instance from a test value (usually a string name).
func New(value interface{}) interface{} {
	name, _ := value.(string)
	return &Node{Name: name, Children: []*Node{}}
}

// Provide a no-op Solution stub so the local runner can compile this file
// even though depthFirstSearch is exercised as a graph/class test.
func Solution(input interface{}) interface{} {
	return nil
}
